Source/Mothership/ThreadComms.cpp

```diff
@@ -0,0 +1,382 @@
+}
+
+/* Takes all messages and pushes them into the queue in the order in which they
+ * are stored in the vector. */
+void ThreadComms::push_MPI_app_queue(std::vector<PMsg_p>* messages)
+{
+    pthread_mutex_lock(&mutex_MPI_app_queue);
+    for (std::vector<PMsg_p>::iterator message = messages->begin();
+         message != messages->end(); message++)
+    {
+        MPIAppQueue.push(*message);
+    }
+    pthread_mutex_unlock(&mutex_MPI_app_queue);
+}
+
+/* ===== BackendOutputQueue ================================================ */
+
+/* If there are packets in the queue, grabs the next packet from
+ * BackendOutputQueue, writes packet with it, and returns true. Otherwise,
+ * returns false. */
+bool ThreadComms::pop_backend_out_queue(std::pair<uint32_t, P_Pkt_t>* packet)
+{
+    bool returnValue = false;
+    pthread_mutex_lock(&mutex_backend_output_queue);
+    if (!BackendOutputQueue.empty())
+    {
+        returnValue = true;
+        *packet = BackendOutputQueue.front();
+        BackendOutputQueue.pop();
+    }
+    pthread_mutex_unlock(&mutex_backend_output_queue);
+    return returnValue;
+}
+
+/* If there are packets in the queue, grabs all packets from
+ * BackendOutputQueue, writes them to packets, and returns true. Otherwise,
+ * returns false. */
+bool ThreadComms::pop_backend_out_queue(
+    std::vector<std::pair<uint32_t, P_Pkt_t> >* packets)
+{
+    bool returnValue = false;
+    packets->clear();
+    pthread_mutex_lock(&mutex_backend_output_queue);
+    while (!BackendOutputQueue.empty())
+    {
+        returnValue = true;
+        packets->push_back(BackendOutputQueue.front());
+        BackendOutputQueue.pop();
+    }
+    pthread_mutex_unlock(&mutex_backend_output_queue);
+    return returnValue;
+}
+
+/* Takes packet and places it into the queue. */
+void ThreadComms::push_backend_out_queue(std::pair<uint32_t, P_Pkt_t> packet)
+{
+    pthread_mutex_lock(&mutex_backend_output_queue);
+    BackendOutputQueue.push(packet);
+    pthread_mutex_unlock(&mutex_backend_output_queue);
+}
+
+/* Takes all packets and pushes them into the queue in the order in which they
+ * are stored in the vector. */
+void ThreadComms::push_backend_out_queue(
+    std::vector<std::pair<uint32_t, P_Pkt_t> >* packets)
+{
+    std::vector<std::pair<uint32_t, P_Pkt_t> >::iterator packet;
+    pthread_mutex_lock(&mutex_backend_output_queue);
+    for (packet = packets->begin(); packet != packets->end(); packet++)
+    {
+        BackendOutputQueue.push(*packet);
+    }
+    pthread_mutex_unlock(&mutex_backend_output_queue);
+}
+
+/* ===== BackendInputQueue ================================================= */
+
+/* Returns true if the number of packets in this queue is too big to fit into
+ * an MPI message given the send buffer size, and false otherwise. */
+bool ThreadComms::is_backend_in_queue_full()
+{
+    /* Be wary - the integer division is intentional. The limit is around 8
+     * million packets per message as of 2020-04-16 on Tinsel.
```

Comment:
So, if I am understanding this correctly from this comment and other comments:

* the thread receiving from the backend will continue to do so until there are no more packets or a message is full?
* The buffer is ~8 million packets

So, hypothetically, if the Supervisor is being spammed by an application (and assuming there is no let up in traffic), the backend will sit there until ~8 million messages have been received before punting it upstream. At ~10,000 packets per second (this is the sustained send rate we see over a link and one device CAN saturate a supervisor), that is about 13 minutes - seems rather high if you ask me... 

:sweet_potato:

Reply:
There's an issue here, but it's a little more nuanced. Remember that the message is "cut loose" as soon as `canRecv()` returns falsy, so a multiple of the 10000 packets-per-second frequency must be in lockstep with the frequency of the `canRecv()` calls that the Mothership makes - otherwise there will be a break in the message.

The way around both of these issues would be to sensibly, artificially limit the number of packets we send in this way. For context, where the current Mothership would hold `8e6` packets before processing, the previous Mothership would hold `1` - some middle ground is needed. This "limit" could be a hardcoded constant, or could be adaptive. I'm a fan of the former (because the behaviour we're trying to catch is quite exceptional, and we can't introspect the sending frequency of an application), but the question remains as to what that number should be.

How long would you be willing to wait before sending a message, bearing in mind that we need to keep the queue as empty as possible? We can't input and output at the same time, because there's no way we can make the queue thread-safe. Given an answer to that question, and a receive rate (remember, 10000 packets per second is not the most common case), we can define the value of such a hardcoded constant.

Well spotted that man :fireworks:.

Reply:
Does it have to be a count-based check? would there be any mileage in making it time-based? e.g. send if you have been buffering for 0.5s, 1s, whatever?

Alternatively, how long are we happy to have messages stall for? What is our latency tolerance?

Reply:
Good question, no idea. I also don't think there's a universal answer, which is why I just wanted to select a constant number of packets and be done with it.

Reply:
Does it make sense to put it in a pre-processor for now and set it to 0.5s (or 1s if you like) worth of messages in the worst case? e.g. assume 10k/s is the worst case, so set the count to 5,000 (or 10,000)?

Reply:
In `ab766a5`, I've defined `BACKEND_QUEUE_PACKET_MAXIMUM` (in `ThreadComms.h`) to `5000`, which I use for the computation in `ThreadComms::is_backend_in_queue_full`. We can naturally keep an eye on this value if we find performance issues, though I'm not expecting any...

Reply:
Also, I opted not to do a time-based check because it would be relatively expensive - we could do a time check after every N packets (as opposed to after every one packet), but I think the packet count will do the job appropriately.

Reply:
:+1: